ddmark/validation_unit_test.go

```diff
@@ -233,6 +233,68 @@ var _ = Describe("Validation Rules Cases", func() {
 			fakeObj.Field2 = 0
 			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(BeNil())
 		})
+
+		It("rejects object with one missing field (Field2 is 3, expected value was 2)", func() {
+			fakeObj.Field2 = 3
+			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(BeNil())
```

Comment:
```suggestion
			Expect(linked.ApplyRule(ValueOf(fakeObj))).ToNot(Succeed())
```

It's more readable to use `Succeed` when function returns an error. Then use `To/ToNo` to express what is expected

In general when you expect an error, I tend to prefer checking the EXACT error message:

```suggestion
			Expect(linked.ApplyRule(ValueOf(fakeObj))).To(MatchError("what is inside the error message"))
```

That will check not only error exists AND is the expected one.

Reply:
I agree it adds a greater level of precision and is better in general.
These error messages are customised to every marker definition because of the included field names -- and I don't like the idea of expecting a string hard-coded error, as it adds toil and I think it provides little added value in this situation.
I'll experiment with a error-generation functions in the DDValidationMarker interface, that would allow to standardise the error generation and reduce the amount of hard-coding to do this. It'll allow for a clearer code base and more precise tests.

Reply:
In the end, I compromised with type-interfaced error generation functions for unit tests, and hard-coded strings for integration tests